internal/controllers/addon/controller.go

```diff
@@ -135,42 +135,39 @@ func (r *AddonReconciler) Reconcile(
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	// Update Addon Status via the kube-api.
+	// This is the only place where the Addon Status is reported back to Kubernetes.
+	defer func() {
+		if updateErr := r.Status().Update(ctx, addon); updateErr != nil {
+			err = updateErr
+		}
+	}()
```

Comment:
Do we really want to update status, even when `err != nil`, when the reconcile function is exiting with an unhandled error?
Depending on where the error is thrown `Addon.Status` might be incomplete or inconsistent.

I would also like to combine this and the `handleUpgradePolicyStatusReporting` defer into one, to make it easier to follow.
Defers execute in reverse order, so right now we would call `handleUpgradePolicyStatusReporting`, before updating the Addon status.
`handleUpgradePolicyStatusReporting` did also call `r.Status().Update()`, which is no longer needed :)

Reply:
> handleUpgradePolicyStatusReporting did also call r.Status().Update(), which is no longer needed :)

Yup, alread handled :D